src/kernel/debug.c

```diff
@@ -0,0 +1,281 @@
+            ksKernelEntry.syscall_no == -SysNBSend ||
+            ksKernelEntry.syscall_no == -SysCall) {
+
+            printf("Cap type: %lu, Invocation tag: %lu\n", (unsigned long) ksKernelEntry.cap_type,
+                   (unsigned long) ksKernelEntry.invocation_tag);
+        }
+        break;
+#ifdef CONFIG_ARCH_ARM
+    case Entry_VCPUFault:
+        printf("VCPUFault\n");
+        break;
+#endif
+#ifdef CONFIG_ARCH_x86
+    case Entry_VMExit:
+        printf("VMExit\n");
+        break;
+#endif
+    default:
+        printf("Unknown\n");
+        break;
+
+    }
+}
+
+void debug_printUserState(void)
+{
+    tcb_t *tptr = NODE_STATE(ksCurThread);
+    printf("Current thread: %s\n", TCB_PTR_DEBUG_PTR(tptr)->tcbName);
+    printf("Next instruction adress: %lx\n", getRestartPC(tptr));
+    printf("Stack:\n");
+    Arch_userStackTrace(tptr);
+}
+
+static const char *string_from_ThreadState(thread_state_t state)
+{
+    switch (thread_state_get_tsType(state)) {
+    case ThreadState_Inactive:
+        return "inactive";
+    case ThreadState_Running:
+        return "running";
+    case ThreadState_Restart:
+        return "restart";
+    case ThreadState_BlockedOnReceive:
+        return "blocked/recv";
+    case ThreadState_BlockedOnSend:
+        return "blocked/send";
+    case ThreadState_BlockedOnReply:
+        return "blocked/reply";
+    case ThreadState_BlockedOnNotification:
+        return "blocked/ntfn";
+#ifdef CONFIG_VTX
+    case ThreadState_RunningVM:
+        return "running VM";
+#endif
+    case ThreadState_IdleThreadState:
+        return "idle";
+    default:
+        break;
+    } /* end switch */
+
+    return "???";
+}
+
+static void debug_dumpSchedulerEx(tcb_t *specific_tcb)
+{
+    /* enough dashes to fill space used by an u64 in hex */
+    static const char *DASHES = "----------------";
+    unsigned int num_tcb = 0;
+
+    if (!specific_tcb) {
+        for (tcb_t *tcb = NODE_STATE(ksDebugTCBs);
+             tcb != NULL;
+             tcb = TCB_PTR_DEBUG_PTR(tcb)->tcbDebugNext) {
+            num_tcb++;
+        }
+        printf("Dump of all TCBs (%u):\n", num_tcb);
+    }
+
+    printf(
+        "  %-*s "
+        "| %-13s "
+        "| %-*s "
+        "| Prio "
+        config_ternary(CONFIG_ENABLE_SMP_SUPPORT, "| Core ", "")
+        config_ternary(CONFIG_KERNEL_MCS, "| RelQ ", "")
+        "| Name"
+        "\n"
+        "  %.*s-" /* dashes for length of SEL4_PRIx_word */
+        "+---------------"
+        "+-%.*s-" /* dashes for length of SEL4_PRIx_word */
+        "+------"
+        config_ternary(CONFIG_ENABLE_SMP_SUPPORT, "+------", "")
+        config_ternary(CONFIG_KERNEL_MCS, "+------", "")
+        "+----------------"
+        "\n",
+        CONFIG_WORD_SIZE / 4, "TCB",
+        "State",
+        CONFIG_WORD_SIZE / 4, "PC",
+        CONFIG_WORD_SIZE / 4, DASHES,
+        CONFIG_WORD_SIZE / 4, DASHES);
+
+    for (tcb_t *tcb = specific_tcb ? specific_tcb : NODE_STATE(ksDebugTCBs);
```

Comment:
I wonder if this should try to dump the scheduler on all cores on day and not just for the current core.